File: DPBO-praktikum/4/cpp/Main.cpp
#include "header.hh"

int main(){
    // membuat beberapa objek mahasiswa
    Mahasiswa mhs1 = Mahasiswa("001","2107922", "Andi", "L","Ilmu Komputer", "FPMIPA", "Universitas Pendidikan Indonesia", "andi@upi.edu");
    Mahasiswa mhs2 = Mahasiswa("002","2107923", "Budi", "L","Bahasa Indonesia" ,"FPEB", "Universitas Pendidikan Indonesia", "budi@upi.edu");
    Mahasiswa mhs3 = Mahasiswa("003","2107924", "Cici", "P", "Psikologi","FIP", "Universitas Pendidikan Indonesia", "cici@upi.edu");
    Mahasiswa mhs4 = Mahasiswa("004","2107925", "Dewi", "P","Psikologi" ,"FIP", "Universitas Pendidikan Indonesia", "dewi@upi.edu");

    // membuat beberapa objek dosen
    Dosen dsn1 = Dosen("001", "123", "Dr. Ahmad", "L", "FPMIPA", "S3", "Artificial Intelligence", "Universitas Pendidikan Indonesia", "ahmad@upi.edu");
    Dosen dsn2 = Dosen("002", "456", "Dr. Roni", "L", "FPEB", "S2", "Ekonomi Pembangunan", "Universitas Pendidikan Indonesia", "budi@upi.edu");
    Dosen dsn3 = Dosen("003", "789", "Prof. Nana", "P", "FIP", "S3", "Psikologi Klinis", "Universitas Pendidikan Indonesia", "cici@upi.edu");

    // membuat beberapa objek course
    Course course1 = Course("Pemrograman Berorientasi Objek", dsn1);
    Course course2 = Course("Ekonomi Mikro", dsn2);
    Course course3 = Course("Psikologi Anak", dsn3);

    // menambahkan mahasiswa ke dalam course
    course1.add_mahasiswa(mhs1);
    course1.add_mahasiswa(mhs2);
    course2.add_mahasiswa(mhs3);
    course3.add_mahasiswa(mhs3);
    course3.add_mahasiswa(mhs4);

    // membuat beberapa objek program studi
    ProgramStudi prodi1 = ProgramStudi("Ilmu Komputer", "ILK");
    ProgramStudi prodi2 = ProgramStudi("Ekonomi", "EKO");
    ProgramStudi prodi3 = ProgramStudi("Psikologi", "PSI");

    // menambahkan course ke dalam program studi
    prodi1.add_course(course1);
    prodi2.add_course(course2);
    prodi3.add_course(course3);

    // membuat vector berisi objek
    vector<Mahasiswa> mahasiswa_list={mhs1, mhs2, mhs3, mhs4};
    vector<Dosen> dosens = {dsn1, dsn2, dsn3};
    vector<Course> courses ={course1, course2, course3};
    vector<ProgramStudi> prodi_list={prodi1, prodi2, prodi3};

    cout << "=======================================" << endl;
    // menampilkan data mahasiswa
    cout << "Daftar Mahasiswa:" << endl;
    for (Mahasiswa mhs : mahasiswa_list) {
    cout << "---------------------------------------" << endl;
    cout << "NIM         : " << mhs.get_nim() << endl;
    cout << "Nama        : " << mhs.get_nama() << endl;
    cout << "Prodi       : " << mhs.get_prodi()<<endl;
    cout << "Fakultas    : " << mhs.get_fakultas() << endl;
    cout << "Universitas : " << mhs.get_asal_universitas() << endl;
    }
    cout << "=======================================" << endl;


    // menampilkan data dosen
    cout << "\n=======================================\n";
    cout << "Daftar Dosen:\n";
    for (Dosen dsn : dosens) {
        cout << "---------------------------------------\n";
        cout << "NIP                 : " << dsn.get_nip() << endl;
        cout << "Nama                : " << dsn.get_nama() << endl;
        cout << "Jenis Kelamin       : " << dsn.get_jenis_kelamin() << endl;
        cout << "Fakultas            : " << dsn.get_fakultas() << endl;
        cout << "Pendidikan Terakhir : " << dsn.get_pend_terakhir() << endl;
        cout << "Keahlian            : " << dsn.get_keahlian() << endl;
    }
    cout << "=======================================\n";

    // menampilkan data course
    cout << "\n=======================================\n";
    cout << "Daftar Course:\n";
    for (Course course : courses ) {
        cout << "---------------------------------------\n";
        cout << "Nama              : " << course.get_nama_matakuliah() << endl;
        cout << "Dosen Pengampu    : " << course.get_dosen().get_nama() << endl;
        cout << "Daftar Mahasiswa  :\n";
        // tampilkan list mahasiswanya
        for (auto mhs : course.get_mahasiswa()) {
            cout << "                   -> " << mhs.get_nama() << endl;
        }
    }
    cout << "=======================================\n";


    // menampilkan data program studi
    cout << "\n=======================================\n";
    cout << "Daftar Program Studi:\n";
    for (auto prodi : prodi_list) {
        cout << "---------------------------------------\n";
        cout << "Nama          : " << prodi.get_nama_prodi() << endl;
        cout << "Kode          : " << prodi.get_kode() << endl;
        cout << "Daftar Course :\n";
        // tampilkan list course nya
        for (auto course : prodi.get_courses()) {
            cout << "               -> " << course.get_nama_matakuliah() << endl;
            cout << "                  Daftar Mahasiswa :\n";
            // tampilkan list mahasiswanya
            for (auto mhs : course.get_mahasiswa()) {
                cout << "                                    --> " << mhs.get_nama() << endl;
            }
        }
    }
    cout << "=======================================\n";

    return 0;
}

File: DPBO-praktikum/4/python/Main.py
# import class yang dibutuhkan
from Mahasiswa import Mahasiswa
from Dosen import Dosen
from Course import Course
from ProgramStudi import ProgramStudi

# membuat beberapa objek mahasiswa
mhs1 = Mahasiswa('2107922', 'Andi', 'L', 'FPMIPA', 'Universitas Pendidikan Indonesia', 'andi@upi.edu')
mhs2 = Mahasiswa('2107923', 'Budi', 'L', 'FPEB', 'Universitas Pendidikan Indonesia', 'budi@upi.edu')
mhs3 = Mahasiswa('2107924', 'Cici', 'P', 'FIP', 'Universitas Pendidikan Indonesia', 'cici@upi.edu')
mhs4 = Mahasiswa('2107925', 'Dewi', 'P', 'FIP', 'Universitas Pendidikan Indonesia', 'dewi@upi.edu')


# membuat beberapa objek dosen
dsn1 = Dosen('001', '123', 'Dr. Ahmad', 'L', 'FPMIPA', 'S3', 'Artificial Intelligence', 'Universitas Pendidikan Indonesia', 'ahmad@upi.edu')
dsn2 = Dosen('002', '456', 'Dr. Roni', 'L', 'FPEB', 'S2', 'Ekonomi Pembangunan', 'Universitas Pendidikan Indonesia', 'budi@upi.edu')
dsn3 = Dosen('003', '789', 'Prof. Nana', 'P', 'FIP', 'S3', 'Psikologi Klinis', 'Universitas Pendidikan Indonesia', 'cici@upi.edu')


# membuat beberapa objek course
course1 = Course('Pemrograman Berorientasi Objek', dsn1)
course2 = Course('Ekonomi Mikro', dsn2)
course3 = Course('Psikologi Anak', dsn3)

# menambahkan mahasiswa ke dalam course
course1.add_mahasiswa(mhs1)
course1.add_mahasiswa(mhs2)
course2.add_mahasiswa(mhs3)
course3.add_mahasiswa(mhs3)
course3.add_mahasiswa(mhs4)

# membuat beberapa objek program studi
prodi1 = ProgramStudi('Ilmu Komputer', 'ILK')
prodi2 = ProgramStudi('Ekonomi', 'EKO')
prodi3 = ProgramStudi('Psikologi', 'PSI')

# menambahkan course ke dalam program studi
prodi1.add_course(course1)
prodi2.add_course(course2)
prodi3.add_course(course3)

print("=======================================")
# menampilkan data mahasiswa
print('Daftar Mahasiswa:')
for mhs in [mhs1, mhs2, mhs3, mhs4]:
    print("---------------------------------------")
    print("NIM         : "+mhs.get_nim())
    print("Nama        : "+mhs.get_nama())
    print("Fakultas    : "+mhs.get_fakultas())
    print("Universitas : "+mhs.get_asal_universitas())
print("=======================================")

# menampilkan data dosen
print("\n=======================================")
print('Daftar Dosen:')
for dsn in [dsn1, dsn2, dsn3]:
    print("---------------------------------------")
    print("NIP                 : "+dsn.get_nip())
    print("Nama                : "+dsn.get_nama())
    print("Jenis Kelamin       : "+dsn.get_jenis_kelamin())
    print("Fakultas            : "+dsn.get_fakultas())
    print("Pendidikan Terakhir : "+dsn.get_pend_terakhir())
    print("Keahlian            : "+dsn.get_keahlian())

print("=======================================")

# menampilkan data course
print("\n=======================================")
print('Daftar Course:')
for course in [course1, course2, course3]:
    print("---------------------------------------")
    print("Nama              : "+course.get_nama_matakuliah())
    print("Dosen Pengampu    : "+course.get_dosen().get_nama())
    print('Daftar Mahasiswa  :')
    # tampilkan list mahasiswanya
    for mhs in course.get_mahasiswa():
        print(f'                   -> {mhs.get_nama()}')
print("=======================================")

# menampilkan data program studi
print("\n=======================================")
print('Daftar Program Studi:')
for prodi in [prodi1, prodi2, prodi3]:
    print("---------------------------------------")
    print("Nama          : "+prodi.get_nama_prodi())
    print("Kode          : "+prodi.get_kode())
    print("Daftar Course : ")
    # tampilkan list course nya
    for course in prodi.get_courses():
        print(f'               -> {course.get_nama_matakuliah()}')
        print("                  Daftar Mahasiswa : ")
        # tampilkan list mahasiswanya
        for mhs in course.get_mahasiswa():
            print(f'                                    --> {mhs.get_nama()}')
print("=======================================")
